### PR TITLE
[SPARK-24360][SQL] Support Hive 3.1 metastore

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Hive 3.1.1 is released. This PR aims to support Hive 3.1.x metastore.
Please note that Hive 3.0.0 Metastore is skipped intentionally.

## How was this patch tested?

Pass the Jenkins with the updated test cases including 3.1.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 1]
```diff
@@ -1179,3 +1180,128 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
 private[client] class Shim_v2_2 extends Shim_v2_1
 
 private[client] class Shim_v2_3 extends Shim_v2_1
+
+private[client] class Shim_v3_1 extends Shim_v2_3 {
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: I think it's fine if the signatures are matched and the tests pass. Let me double check them within tomorrow.

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Thanks!

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 2]
```diff
@@ -100,6 +100,7 @@ private[hive] object IsolatedClientLoader extends Logging {
     case "2.1" | "2.1.0" | "2.1.1" => hive.v2_1
     case "2.2" | "2.2.0" => hive.v2_2
     case "2.3" | "2.3.0" | "2.3.1" | "2.3.2" | "2.3.3" | "2.3.4" => hive.v2_3
+    case "3.1" | "3.1.0" | "3.1.1" => hive.v3_1
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: What is the reason we do not support 3.0?

Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: It's not stable like Hadoop 3.0. We skip Hadoop 3.0 and go with Hadoop 3.1.

**Comment by dongjoon-hyun** [comment]
Also, cc @wangyum since he is working on Hive upgrade.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Then, hopefully, I can proceed that in another PR.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 3]
```diff
@@ -1179,3 +1180,128 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
...
+      clazzLoadFileType.getEnumConstants.find(_.toString.equalsIgnoreCase("REPLACE_ALL"))
+    } else {
+      clazzLoadFileType.getEnumConstants.find(_.toString.equalsIgnoreCase("KEEP_EXISTING"))
+    }
+    assert(loadFileType.isDefined)
+    loadDynamicPartitionsMethod.invoke(hive, loadPath, tableName, partSpec, loadFileType.get,
+      numDP: JInteger, listBucketingLevel, isAcid, writeIdInLoadTableOrPartition,
+      stmtIdInLoadTableOrPartition, hasFollowingStatsTask, AcidUtils.Operation.NOT_ACID,
+      replace: JBoolean)
+  }
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: It's possible, but it doesn't help much. As we see [here](https://github.com/apache/spark/pull/23694/files#diff-6fd847124f8eae45ba2de1cf7d6296feL855), `HiveClient.getIndexes` raises `NoSuchMethodError` before we call `shim.dropIndex`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 4]
```diff
@@ -1179,3 +1180,128 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
 private[client] class Shim_v2_2 extends Shim_v2_1
 
 private[client] class Shim_v2_3 extends Shim_v2_1
+
+private[client] class Shim_v3_1 extends Shim_v2_3 {
+  // Spark supports only non-ACID operations
+  protected lazy val isAcidIUDoperation = JBoolean.FALSE
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: @dongjoon-hyun, isn't it `isAcid` defined at `Shim_v0_14`? Was wondering why this was separate variable again. Do you see any possibility that this is different specifically for 3.1? Then, it's fine.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 5]
```diff
@@ -1179,3 +1180,128 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
...
+      clazzLoadFileType.getEnumConstants.find(_.toString.equalsIgnoreCase("KEEP_EXISTING"))
+    }
+    assert(loadFileType.isDefined)
+    loadPartitionMethod.invoke(hive, loadPath, table, partSpec, loadFileType.get,
+      inheritTableSpecs: JBoolean, isSkewedStoreAsSubdir: JBoolean,
+      isSrcLocal: JBoolean, isAcid, hasFollowingStatsTask,
+      writeIdInLoadTableOrPartition, stmtIdInLoadTableOrPartition, replace: JBoolean)
+  }
+
+  override def loadTable(
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: checked

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 6]
```diff
@@ -1179,3 +1180,128 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
...
+      tableName: String,
+      replace: Boolean,
+      isSrcLocal: Boolean): Unit = {
+    val loadFileType = if (replace) {
+      clazzLoadFileType.getEnumConstants.find(_.toString.equalsIgnoreCase("REPLACE_ALL"))
+    } else {
+      clazzLoadFileType.getEnumConstants.find(_.toString.equalsIgnoreCase("KEEP_EXISTING"))
+    }
+    assert(loadFileType.isDefined)
+    loadTableMethod.invoke(hive, loadPath, tableName, loadFileType.get, isSrcLocal: JBoolean,
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: checked

**Review by HyukjinKwon — APPROVED** [review]
LGTM except one question

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Yea, I noticed it. I am leaving this comment orthogonally just as a sanity check.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Historically, `isAcidIUDoperation` is an evolved one from `isAcid`.

In Hive code, `isAcid` was a general ACID operation and `isAcidIUDoperation` is now used for ACID Insert/Update/Delete operations. Also, they checks `isFullAcidTable` and use it together like this.
```java
else if(!isAcidIUDoperation && isFullAcidTable) {
    destPath = fixFullAcidPathForLoadData(loadFileType, destPath, txnId, stmtId, tbl);
}
```

We don't know the future of Hive. So, for the different parameter name, we had better handle differently. That was my logic.

**Pull request closed without merging**